### PR TITLE
feat(header): cache hash

### LINKED ISSUE #2959: header.go: `Hash()` should recompute / read from cached value of actually computed hash of header
Rather than taking it from the commit BlockID.

Also implement caching inside the header struct on computation.

### PR DESCRIPTION
**Pull request by cristaloleg**
Closes #2959 and Closes #3189

### CODE REVIEW

File: header/header.go [hunk 1]
```diff
@@ -94,7 +97,22 @@ func (eh *ExtendedHeader) Time() time.Time {
 // NOTE: It purposely overrides Hash method of RawHeader to get it directly from Commit without
 // recomputing.
```
Comment (cristaloleg) [inline header/header.go:98]: Well, practically we're recomputing it but on the other side we do this when it's not present. Any ideas for the best possible doc-comment here? (or just drop the part about recomputing, lol)

Comment (Wondertan) [reply header/header.go:98]: And added a fresh one

File: header/header.go [hunk 2]
```diff
@@ -94,7 +97,22 @@ func (eh *ExtendedHeader) Time() time.Time {
...
 func (eh *ExtendedHeader) Hash() libhead.Hash {
-	return eh.Commit.BlockID.Hash.Bytes()
+	if len(eh.hash) == 0 {
+		bs := make([][]byte, len(eh.Commit.Signatures))
+		for i, commitSig := range eh.Commit.Signatures {
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply header/header.go:102]: Commit.Hash != BlockID.Hash

[the remark this answers is not in the document]
Comment (Wondertan) [reply header/header.go:102]: So we should do `eh.RawHeader.Hash()` + hashing

[the remark this answers is not in the document]
Comment (cristaloleg) [reply header/header.go]: I blindly assumed that you approved previous PR, thankfully tests confirm that.

File: header/header.go [hunk 3]
```diff
@@ -38,6 +39,8 @@ type ExtendedHeader struct {
 	Commit       *core.Commit               `json:"commit"`
 	ValidatorSet *core.ValidatorSet         `json:"validator_set"`
 	DAH          *da.DataAvailabilityHeader `json:"dah"`
+
+	hash []byte
```
[the remark this answers is not in the document]
Comment (cristaloleg) [reply header/header.go]: Fixed with `sync.Once.`

File: header/header.go (hunk 1 above)
Comment (cristaloleg) [reply header/header.go]: Done

File: header/header.go [hunk 4]
```diff
@@ -223,7 +229,7 @@ func (eh *ExtendedHeader) UnmarshalBinary(data []byte) error {
 		return err
 	}
 
-	*eh = *out
+	*eh = *out.Clone()
```
Comment (Wondertan) [inline header/header.go:232]: Why is this necessary?

Comment (cristaloleg) [reply header/header.go:232]: 'cause now we've `sync.Once` that cannot (should not) be copied. We've a few places where we do assignment which triggers `go vet`.

So, additional helper method to prevent such things. See failure: https://github.com/celestiaorg/celestia-node/actions/runs/12118920400/job/33784451925?pr=3975#step:4:34

Comment (Wondertan) [reply header/header.go:232]: Can we find a way to solve this without introducing additional copy?

Comment (cristaloleg) [reply header/header.go]: Well, ok, done.

File: header/header.go [hunk 5]
```diff
@@ -223,7 +229,12 @@ func (eh *ExtendedHeader) UnmarshalBinary(data []byte) error {
 		return err
 	}
 
-	*eh = *out
+	*eh = ExtendedHeader{
+		RawHeader:    out.RawHeader,
+		Commit:       out.Commit,
+		ValidatorSet: out.ValidatorSet,
+		DAH:          out.DAH,
+	}
```
Comment (Wondertan) [inline header/header.go:237]: Isn't this an equivalent of what it was before, but without a method?

**Review by Wondertan — COMMENTED** [review]
I carefully reviewed this again and realized that we actually don't need this change.

[Here](https://github.com/celestiaorg/celestia-node/blob/d726f1131c89536d967f754eb48b49e1cd0aa2f6/header/header.go#L150) we check that the result of local hashing is equal to what we have in the commit already. So when we return from commit we know its something we computed ourselves. Therefore, we keep returning hash from the commit and there is no need to add recompute value and synchronization around it.

File: header/header.go (hunk 5 above)
Comment (Wondertan) [reply header/header.go:237]: (not relevant)

Comment (cristaloleg) [reply header/header.go:237]: Yes, it's. We did shallow copy before.

**Comment by cristaloleg** [comment]
Closing as not needed (see Hlib's review comment)
